db/crud.go

```diff
@@ -882,278 +1072,109 @@ func (db *Database) IsIllegalConflict(doc *Document, parentRevID string, deleted
-		body, newAttachments, updatedExpiry, err = callback(doc)
-		if err != nil {
-			return
-		}
+	err = db.postCallbackValidation(body)
+	if err != nil {
+		return doc, syncFunctionExpiry, body, ``, storedBody, ``, previousDocSequence, unusedSequences, changedPrincipals, changedRoleUsers, err
+	}
 
-		//Reject a body that contains the "_removed" property, this means that the user
-		//is trying to update a document they do not have read access to.
-		if body["_removed"] != nil {
-			err = base.HTTPErrorf(http.StatusNotFound, "Document revision is not accessible")
-			return
-		}
+	newRevID := body[BodyRev].(string)
+	prevCurrentRev := doc.CurrentRev
+	doc.updateWinningRevAndSetDocFlags()
+	storedBody = db.storeOldBodyInRevTreeAndUpdateCurrent(doc, prevCurrentRev, newRevID, body)
 
-		//Reject bodies containing user special properties for compatibility with CouchDB
-		if containsUserSpecialProperties(body) {
-			err = base.HTTPErrorf(400, "user defined top level properties beginning with '_' are not allowed in document body")
-			return
-		}
+	body[BodyId] = doc.ID
+	syncExpiry, oldBodyJSON, channelSet, access, roles, err := db.runSyncFn(doc, body, newRevID)
+	if err != nil {
+		return doc, syncFunctionExpiry, body, newRevID, storedBody, oldBodyJSON, previousDocSequence, unusedSequences, changedPrincipals, changedRoleUsers, err
+	}
 
-		// Determine which is the current "winning" revision (it's not necessarily the new one):
-		newRevID = body[BodyRev].(string)
-		prevCurrentRev := doc.CurrentRev
-		var branched, inConflict bool
-		doc.CurrentRev, branched, inConflict = doc.History.winningRevision()
-		doc.setFlag(channels.Deleted, doc.History[doc.CurrentRev].Deleted)
-		doc.setFlag(channels.Conflict, inConflict)
-		doc.setFlag(channels.Branched, branched)
+	if len(channelSet) > 0 {
+		doc.History[newRevID].Channels = channelSet
+	}
 
-		// If tombstone, write tombstone time
-		if doc.hasFlag(channels.Deleted) {
-			doc.SyncData.TombstonedAt = time.Now().Unix()
-		} else {
-			doc.SyncData.TombstonedAt = 0
-		}
+	err = db.addAttachments(newAttachments)
+	if err != nil {
+		return doc, syncFunctionExpiry, body, newRevID, storedBody, oldBodyJSON, previousDocSequence, unusedSequences, changedPrincipals, changedRoleUsers, err
+	}
 
-		if doc.CurrentRev != prevCurrentRev && prevCurrentRev != "" && doc.Body() != nil {
-			// Store the doc's previous body into the revision tree:
-			bodyJSON, marshalErr := doc.MarshalBody()
-			if marshalErr != nil {
-				base.WarnfCtx(db.Ctx, base.KeyAll, "Unable to marshal document body for storage in rev tree: %v", marshalErr)
-			}
-			doc.setNonWinningRevisionBody(prevCurrentRev, bodyJSON, db.AllowExternalRevBodyStorage())
-		}
+	db.backupAncestorRevs(doc, newRevID, storedBody)
 
-		// Store the new revision body into the doc:
-		storedBody = doc.setRevisionBody(newRevID, body, db.AllowExternalRevBodyStorage())
+	unusedSequences, previousDocSequence, err = db.handleSequences(previousDocSequence, doc, unusedSequences)
```

Comment:
It's misleading that this is setting 'previousDocSequence', since what's being returned is actually the current doc sequence.  I suggest modifying handleSequences to no longer return this value at all, since it's already available as doc.Sequence.  Then modify line 1153 to return doc.Sequence.

Reply:
For similar reasons documentUpdateFunc doesn't actually need to return retDocSequence (since it's available via retDoc) - it might make the intent clearer to have the callers manage previous docSequence (see comment there).